src/apps/iq/ethereum/iq.hiiq.contract-position-fetcher.ts

```diff
@@ -0,0 +1,60 @@
+import { Inject } from '@nestjs/common';
+import { ethers } from 'ethers';
+
+import { IAppToolkit, APP_TOOLKIT } from '~app-toolkit/app-toolkit.interface';
+import { PositionTemplate } from '~app-toolkit/decorators/position-template.decorator';
+import { MetaType } from '~position/position.interface';
+import { ContractPositionTemplatePositionFetcher } from '~position/template/contract-position.template.position-fetcher';
+import { GetDataPropsParams, GetTokenBalancesParams } from '~position/template/contract-position.template.types';
+
+import { IqContractFactory, IqHiiq } from '../contracts';
+
+@PositionTemplate()
+export class EthereumIqHiiqContractPositionFetcher extends ContractPositionTemplatePositionFetcher<IqHiiq> {
+  groupLabel = 'Locked';
+
+  constructor(
+    @Inject(APP_TOOLKIT) protected readonly appToolkit: IAppToolkit,
+    @Inject(IqContractFactory) protected readonly iqContractFactory: IqContractFactory,
+  ) {
+    super(appToolkit);
+  }
+
+  getContract(address: string): IqHiiq {
+    return this.iqContractFactory.iqHiiq({ address, network: this.network });
+  }
+
+  async getDefinitions() {
+    return [{ address: '0xb55Dcc69d909103b4De773412A22AB8B86e8c602' }];
+  }
+
+  async getTokenDefinitions() {
+    return [
+      {
+        metaType: MetaType.LOCKED,
+        address: '0x579cea1889991f68acc35ff5c3dd0621ff29b0c9',
+        network: this.network,
+      },
+    ];
+  }
+
+  async getLabel() {
+    return `HiIQ Lock`;
+  }
+
+  async getDataProps(_params: GetDataPropsParams<IqHiiq>) {
+    const defaultDataProps = await super.getDataProps(_params);
+    const { contract } = _params;
+    const totalHiIQ = await contract.totalHiIQSupplyStored();
+    const totalHiIQSupply = ethers.utils.formatUnits(totalHiIQ);
+    return { ...defaultDataProps, totalHiIQSupply };
+  }
+
+  async getTokenBalancesPerPosition({ address, contract }: GetTokenBalancesParams<IqHiiq>) {
+    const bigTime = await contract.userHiIQEndpointCheckpointed(address);
+    let timeRemaining = bigTime.toNumber();
+    const unlockTime = new Date(timeRemaining * 1000).toUTCString();
+    const earned = await contract.earned(address);
+    return [earned, unlockTime];
```

Comment:
If you want to add extra informations about the position like `unlocktime` It should be added to the dataProps instead

Reply:
Okay will do

Reply:
But what about the address param that i need.
like i need the address of the user.
do i do it like this?

  async getDataProps(_params: GetDataPropsParams<IqHiiq>) {
    const defaultDataProps = await super.getDataProps(_params);
    const { address, contract } = _params; // the address here meaning the address of each user or does it mean the address of the contract